### PR TITLE
Update README.lws_sul.md - described 'lws_container_of', added note about 'lws_sorted_usec_list_t' zeroing

### PR DESCRIPTION
**Pull request by inobelar**
- Described `lws_container_of(...)` usage
- Added note about `lws_sorted_usec_list_t` zeroing.
- Added `c` tags to enable highlighting in markdown code blocks

The main reason for it - bug in my own code, caused by missing `lws_sorted_usec_list_t` zero-initialization.

In the next (pseudo)-code (pretty simple/common, but listed here to to show how bug may be happened):

```c
/* Header: Public interface */

typedef struct websockets_events_receiver websockets_events_receiver_t;

websockets_events_receiver_t* websockets_events_receiver_init(
    const char* client_id,
    const char* address, int port,
    int reconnect_interval_ms
);

void websockets_events_receiver_deinit(websockets_events_receiver_t* self);

/* Other methods for ... taking some custom data from receiver :D */
```

```c
/* Source: Implementation */

#include <websockets_events_receiver/websockets_events_receiver.h>

#include <pthread.h>

#include <libwebsockets.h>

#include <stdio.h>  /* for: fprintf(), fflush(), sprintf() */
#include <stdlib.h> /* for: alloc(), free() */
#include <string.h> /* for: memcpy() */

#include <... a lot of other includes ...>

/*
    References:
        https://github.com/iamscottmoyers/simple-libwebsockets-example/blob/master/client.c

        https://github.com/warmcat/libwebsockets/blob/main/minimal-examples-lowlevel/ws-client/minimal-ws-client-ping/minimal-ws-client-ping.c
        https://github.com/warmcat/libwebsockets/blob/main/minimal-examples-lowlevel/ws-client/minimal-ws-client-rx/minimal-ws-client.c

        https://git.rwth-aachen.de/acs/public/villas/node/-/blob/f77231f2a76dea3493076a930d73510f94cf6689/lib/nodes/webrtc/signaling_client.cpp

        https://github.com/warmcat/libwebsockets/issues/1612
*/

/* -------------------------------------------------------------------------- */

struct websockets_events_receiver
{
    char* client_id;

    char* address;
    int port;

    int reconnect_interval_ms;

    /* -------------------------------------------------- */

    pthread_t thread_handle;
    /*bool*/ int thread_valid;

    /*bool*/ int is_thread_running;

    /* -------------------------------------------------- */

    struct lws_context* websockets_context;

    struct lws* websockets_wsi; /* WebSockets Interface */

    lws_sorted_usec_list_t websockets_sul; /* Scheduler handle */
};

/* -------------------------------------------------------------------------- */

static int
protocol_callback(
    struct lws *wsi, enum lws_callback_reasons reason,
    void *user,
    void *in, size_t len
);

/* The registered protocols */
static const struct lws_protocols protocols[] =
{
    {
        .name                  = "remote_events",    /* Protocol name*/
        .callback              = protocol_callback,  /* Protocol callback */
        .per_session_data_size = 0,                  /* Protocol callback 'userdata' size */
        .rx_buffer_size        = 1024,               /* Receve buffer size (0 = no restriction) */
        .id                    = 0,                  /* Protocol Id (version) (optional) */
        .user                  = NULL,               /* 'User data' ptr, to access in 'protocol callback */
        .tx_packet_size        = 0                   /* Transmission buffer size restriction (0 = no restriction) */
    },
    LWS_PROTOCOL_LIST_TERM /* terminator */
};

static void
connect_callback(lws_sorted_usec_list_t *sul)
{
    websockets_events_receiver_t* self = lws_container_of(sul, websockets_events_receiver_t, websockets_sul);

    debug_printf("Trying to connect (client_id=%s) ...\n", self->client_id);

    struct lws_client_connect_info connect_info;
    {
        memset(&connect_info, 0, sizeof(struct lws_client_connect_info));

        connect_info.context             = self->websockets_context;
        connect_info.address             = self->address;
        connect_info.port                = self->port;
        connect_info.path                = "/";
        connect_info.host                = self->address;
        connect_info.origin              = self->address;
        /*connect_info.ssl_connection = !strcmp(protocol, "https"); */
        connect_info.protocol            = protocols[0].name;
        connect_info.local_protocol_name = protocols[0].name; /* optional */
        connect_info.pwsi                = &(self->websockets_wsi); /* Set this websocket to be 'parent socket' to newly created - to reuse it */
        connect_info.opaque_user_data    = self; /* Set 'user data', that can be given in 'protocol callback' via: `lws_get_opaque_user_data(struct lws* wsi)` */
    }

    /* Try to connect to server */
    self->websockets_wsi = lws_client_connect_via_info(&connect_info);

    if(self->websockets_wsi == NULL) /* Connection attempt failed */
    {
        if(self->reconnect_interval_ms >= 0) /* Reconnection enabled */
        {
            /* Schedule reconnection attempt after 'reconnect interval' milliseconds */
            lws_sul_schedule(self->websockets_context, 0, &(self->websockets_sul), connect_callback, self->reconnect_interval_ms * LWS_US_PER_MS);
        }
        else /* Reconnect interval < 0 --> Reconnection disabled */
        {
            self->is_thread_running = /*false*/ 0;
        }
    }
}

/*
    NOTE: in 'protocol callbacks' resulting return 'int' status means:
    - zero     - keep connection opened
    - non zero - close connection
*/
static int
protocol_callback(
    struct lws *wsi, enum lws_callback_reasons reason,
    void *user,
    void *in, size_t len
)
{
    /*
        Get 'user data', assigned in `lws_client_connect_info::opaque_user_data`
        Reference: https://stackoverflow.com/a/75014998/
     */
    websockets_events_receiver_t* self = lws_get_opaque_user_data(wsi);

    /* For which reason was this callback called? */
    switch(reason)
    {
        /* The connection was successfully established */
        case LWS_CALLBACK_CLIENT_ESTABLISHED:
        {
            /* Request 'writing' - for sending first messages */
            lws_callback_on_writable(wsi);
        } break;

        /* There was an error connecting to the server */
        case LWS_CALLBACK_CLIENT_CONNECTION_ERROR:
        {
            debug_printf("<!> Client connection error (client_id=%s): %s\n", self->client_id, ((in != NULL) ? (char*)in : "(null)") );

            goto connection_error;
        } break;

        /* The connection closed */
        case LWS_CALLBACK_CLIENT_CLOSED:
        {
            goto connection_error;
        } break;

        /* The server notifies us that we can write data */
        case LWS_CALLBACK_CLIENT_WRITEABLE:
        {
            /*
                ...
                Here we send message about ourself, etc
                ...
            */
        } break;

        /* Our client received something */
        case LWS_CALLBACK_CLIENT_RECEIVE:
        {
            /* 
                ...
                Here we parse received data into 'remote events' structures, lock mutex, add events to queue, etc 
                ...
            */
        } break;

        default:
        {
            /* NOOP */
        } break;

    }

    return lws_callback_http_dummy(wsi, reason, user, in, len);

    connection_error:
        if(self->reconnect_interval_ms >= 0) /* Reconnection enabled */
        {
            /* Schedule reconnection attempt after 'reconnect interval' milliseconds */
            lws_sul_schedule(self->websockets_context, 0, &(self->websockets_sul), connect_callback, self->reconnect_interval_ms * LWS_US_PER_MS);
        }
        else /* Reconnect interval < 0 --> Reconnection disabled */
        {
            self->is_thread_running = /*false*/ 0;
        }

        /*
            Here is we return '0' instead 'non-zero' (that indicates that we
            need to close connection), since it already happened for connection,
            that closed.
         */
        return 0;
}

/* -------------------------------------------------------------------------- */

static void*
thread_func(void* arg)
{
    websockets_events_receiver_t* self = (websockets_events_receiver_t*)(arg);

    struct lws_context_creation_info context_creation_info;
    {
        memset(&context_creation_info, 0, sizeof(struct lws_context_creation_info));

        context_creation_info.port = CONTEXT_PORT_NO_LISTEN; /* We do not run any server */
        context_creation_info.protocols = protocols;

        context_creation_info.gid = -1;
        context_creation_info.uid = -1;
    }

    self->websockets_context = lws_create_context(&context_creation_info);
    if(self->websockets_context == NULL)
    {
        return NULL;
    }

    /* ---------------------------------------------------------------------- */

    self->is_thread_running = /*true*/ 1;

    /* Try to connect to server (immediately) */
    lws_sul_schedule(self->websockets_context, 0, &(self->websockets_sul), connect_callback, 1);

    int n = 0;
    while( (self != NULL) && (self->is_thread_running == /*true*/ 1) && (n >= 0) )
    {
        n = lws_service(self->websockets_context, /* timeout_ms = */ 250 ); /* NOTE: since v3.2, timeout_ms may be set to '0', since it internally ignored */
    }

    /* Stop sheduler (cancel scheduled callbacks) */
    lws_sul_schedule(self->websockets_context, 0, &(self->websockets_sul), NULL, LWS_SET_TIMER_USEC_CANCEL); /* Same as: 'lws_sul_cancel()' */

    /* ---------------------------------------------------------------------- */

    lws_context_destroy(self->websockets_context);
    self->websockets_context = NULL;

    return NULL;
}

websockets_events_receiver_t* websockets_events_receiver_init(
    const char* client_id,
    const char* address, int port,
    int reconnect_interval_ms
)
{
    websockets_events_receiver_t* self = (websockets_events_receiver_t*) malloc( sizeof(websockets_events_receiver_t) );
    if(self == NULL)
    {
        return NULL;
    }

    /* ------------------------------------------------- */
    /* Initialization (by default - null) */

    self->client_id              = NULL;
    self->address                = NULL;
    self->port                   = 0;
    self->reconnect_interval_ms  = 0;

    self->thread_handle          = 0; /* 'zeroing' here is optional */
    self->thread_valid           = /*false*/ 0;
    self->is_thread_running      = /*false*/ 0;

    self->websockets_context     = NULL;
    self->websockets_wsi         = NULL;

    /* ---------------------------------------------------------------------- */

    /* client id */
    {
        const size_t length = strlen(client_id);
        if(length > 0)
        {
            self->client_id = (char*) malloc(length + 1);
            if(self->client_id != NULL)
            {
                memcpy(self->client_id, client_id, length);
                self->client_id[length] = '\0';
            }
        }
        else
        {
            self->client_id = NULL;
        }
    }

    /* address */
    {
        const size_t length = strlen(address);
        if(length > 0)
        {
            self->address = (char*) malloc(length + 1);
            if(self->address != NULL)
            {
                memcpy(self->address, address, length);
                self->address[length] = '\0';
            }
        }
        else
        {
            self->address = NULL;
        }
    }

    self->port = port;

    self->reconnect_interval_ms = reconnect_interval_ms;

    /* ------------------------------------------------- */

    pthread_attr_t thread_attributes;
    pthread_attr_init(&thread_attributes);
    pthread_attr_setdetachstate(&thread_attributes, PTHREAD_CREATE_JOINABLE);

    const int thread_create_status =
        pthread_create(&(self->thread_handle), &thread_attributes, &thread_func, /*arg*/ self);

    pthread_attr_destroy(&thread_attributes);

    self->thread_valid = (thread_create_status == 0) ? /*true*/ 1 : /*false*/ 0;

    if(thread_create_status != 0)
    {
        goto initialization_failed;
    }

    /* ---------------------------------------------------------------------- */

    return self;

    initialization_failed:
        websockets_events_receiver_deinit(self);
        self = NULL;

        return NULL;
}

void websockets_events_receiver_deinit(websockets_events_receiver_t* self)
{
    if(self != NULL)
    {
        self->is_thread_running = /*false*/ 0; /* Set flag, that we need to stop the thread */

        /* Stop 'lws_service()' */
        if(self->websockets_context != NULL)
        {
            lws_cancel_service(self->websockets_context);
        }

        if(self->thread_valid == /*true*/ 1)
        {
            void* join_result = NULL;
            const int join_status = pthread_join(self->thread_handle, &join_result);
            (void) join_result; /* Unused */

            (void) join_status; /* Unused */ /* TODO: check it */

            self->thread_valid = /*false*/ 0;
        }

        if(self->address != NULL)
        {
            free(self->address);

            self->address = NULL;
        }

        if(self->client_id != NULL)
        {
            free(self->client_id);

            self->client_id = NULL;
        }

        free(self);
    }
}
```

`websockets_events_receiver self` was allocated by `malloc()`, but was not entirely zeroed by `memcpy()`, so ` lws_sorted_usec_list_t websockets_sul;` contains garbage, and produce `segmentation fault`, when 
`lws_sul_schedule(self->websockets_context, 0, &(self->websockets_sul), connect_callback, 1);` 
called. 

GDB backtrace show the next:

```
lws_dll2_remove (d=0x555558fe59b8) at /libwebsockets-main/lib/core/lws_dll2.c:209
209         d->next->prev = d->prev;
(gdb) backtrace
#0  lws_dll2_remove (d=0x555558fe59b8)
    at /libwebsockets-main/lib/core/lws_dll2.c:209
#1  0x0000555555d3bb9a in __lws_sul_insert
    (own=0x7ffff0000e90, sul=0x555558fe59b8)
    at /libwebsockets-main/lib/core-net/sorted-usec-list.c:53
#2  0x0000555555d3bcc8 in lws_sul2_schedule
    (context=0x7ffff0000b70, tsi=0, flags=0, sul=0x555558fe59b8)
    at /libwebsockets-main/lib/core-net/sorted-usec-list.c:86
#3  0x0000555555d3c11e in lws_sul_schedule
    (ctx=0x7ffff0000b70, tsi=0, sul=0x555558fe59b8, _cb=0x555555d18870 <connect_callback>, _us=1)
    at /libwebsockets-main/lib/core-net/sorted-usec-list.c:292
#4  0x0000555555d18f8f in thread_func (arg=0x555558fe57a0)
```

----

**TL;DR**: I think its important to notify users in docs about zeroing :)

### CODE REVIEW

**Comment by lws-team** [comment]
Thanks, it's pushed on main + v.43-stable.

**Comment by inobelar** [comment]
@lws-team note: https://libwebsockets.org/lws-api-doc-main/html/md_READMEs_README_8lws_sul.html still not updated, it looks like site docs not re-generated automatically.
